src/workflows/recipes/components/PreferenceRolloutFields.js

```diff
@@ -0,0 +1,295 @@
+      };
+    });
+  };
+
+  render() {
+    const { disabled, recipeArguments } = this.props;
+
+    return (
+      <div>
+        <Row>
+          <p className="action-info">Roll out a permanent preference change.</p>
+          <Col sm={24} md={11}>
+            <FormItem
+              label="Slug"
+              name="arguments.slug"
+              initialValue={recipeArguments.get('slug', '')}
+            >
+              <Input disabled={disabled} />
+            </FormItem>
+          </Col>
+        </Row>
+
+        <p className="action-info">Preferences</p>
+
+        {this.state.rows.map(row => (
+          <RowField
+            row={row}
+            defaultValues={DEFAULT_VALUES}
+            disabled={disabled}
+            key={row.id}
+            disableRemove={this.state.rows.size <= 1}
+            onRowChange={row => {
+              this.changeRow(row);
+            }}
+            onRowRemove={row => {
+              this.removeRow(row);
+            }}
+          />
+        ))}
+
+        <Button
+          disabled={disabled}
+          type="default"
+          icon="plus"
+          onClick={this.handleClickAddNewRow}
+          id="add-row"
+        >
+          Add Row
+        </Button>
+      </div>
+    );
+  }
+}
+
+export default PreferenceRolloutFields;
+
+export class RowField extends React.PureComponent {
+  state = {
+    name: this.props.row.name,
+    type: this.props.row.type,
+    value: this.props.row.value,
+  };
+
+  static propTypes = {
+    row: PropTypes.object.isRequired,
+    defaultValues: PropTypes.object,
+    disabled: PropTypes.bool,
+    disableRemove: PropTypes.bool,
+  };
+
+  static defaultProps = {
+    defaultValues: DEFAULT_VALUES,
+    disabled: false,
+    disableRemove: false,
+  };
+
+  _bubbleUpRow = () => {
+    this.props.onRowChange({
+      name: this.state.name,
+      type: this.state.type,
+      value: this.state.value,
+    });
+  };
+
+  handleNameChange = name => {
+    this.setState(() => {
+      return { name: name.trim() };
+    }, this._bubbleUpRow);
+  };
+
+  handleTypeChange = type => {
+    const { defaultValues } = this.props;
+    if (!defaultValues.hasOwnProperty(type)) {
+      throw new Error(`Unrecognized value type for '${type}'`);
+    }
+    this.setState(
+      {
+        type,
+        value: defaultValues[type],
+      },
+      this._bubbleUpRow,
+    );
+  };
+
+  handleValueChange = value => {
+    this.setState({ value }, this._bubbleUpRow);
+  };
+
+  render() {
+    // The reason for passing the `i` is so that we can have a unique thing that
+    // makes the `name` on each <FormItem> different and unique.
+    const { disabled, row, disableRemove } = this.props;
+    const { id } = row;
+    const { name, type, value } = this.state;
+
+    let valueInput;
+    if (type === 'string') {
+      valueInput = (
+        <Input
+          disabled={disabled}
+          onChange={event => this.handleValueChange(event.target.value)}
```

Comment:
+1 for event handlers using the event object rather than things derived from it.

Reply:
> is bad for performance

Not necessarily. 
See https://cdb.reacttraining.com/react-inline-functions-and-performance-bdff784f5578
and https://twitter.com/ryanflorence/status/905574498400845825?lang=en and https://twitter.com/ryanflorence/status/911023769765871616

I *like* having inline functions because they're local. Meaning, their functionality is there right near where it matters and gets used. If you use public field methods or bound functions you have to scroll away from where it's used. 

Sometimes it's nice to break out into bound functions (or better, public field methods) to avoid DRY. E.g.

```javascript
class C extends React.Component {
  reload = () => { window.location.reload(true) }
  render() {
    return <div>
      <button onClick={this.reload}>Reload</button>
      ...
      <button onClick={this.reload}>Reload</button>
    </div>
  } 
}
```
is arguably better than:
```javascript
class C extends React.Component {
  render() {
    return <div>
      <button onClick={() => { window.location.reload(true) }}>Reload</button>
      ...
      <button onClick={() => { window.location.reload(true) }}>Reload</button>
    </div>
  } 
}
```

Also, the [Input](https://github.com/ant-design/ant-design/blob/32c78b2f8dd7f3b3ec059e956f07f083417259de/components/input/Input.tsx#L27) component is based on `React.Component` without a `shouldComponentUpdate` so it doesn't even bother doing a diff. It just goes straight to render. React's virtual DOM will concern itself whether to update the DOM or not and that's usually the genuinely expensive thing. 

As a matter of fact; we should probably reconsider our heavy usage of `React.PureComponent` anyway. As Ryan pointed out in his blog post, he benchmarked his app before he added `PureRenderMixin` (what it was before `React.PureComponent` was part of the core) and after. 
> "Much to my surprise, my app got slower 🤔."

That's because the majority of the time, the props (or state) were different and most of the time needed to re-render anyway so the shallow diffs that `React.PureComponent` had to do were just costing time. 

The real point of Ryan's blog post is that "premature optimization is evil" and that real optimization is measurements. There are probably enormously larger fish to fry in areas of the code that gets used a heck lot more. 

**In conclusion...** 

* Just because a `render` function executes doesn't mean it's bad for performance. What matters is updates to the DOM.
* Doing diffs like `React.PureComponent` does, might cost more than it saves in avoidances of calling `render`.
* I prefer the style of inline functions because it makes the code nicer to read often. I prefer to refactor that out to public field methods to avoid DRY or if feel there's some other reason to do so for clarity.
* Performance isn't about gut feelings but about measurements.
* Even if there's a small performance difference, in either direction, this deep in the component hierarchy, it really doesn't matter that much.

Reply:
Now I remember why I did it the way I did it. 
I wanted to make my public field methods disconnected from the minutea of inputs and event handling stuff. The field methods `handleValueChange`, `handleNameChange`, and `handleTypeChange` gets the real data. Not events. 

The `onChange` prop on components like `SwitchBox` and `InputNumber` (both from `antd`) don't work like your regular pure React `<input type="text" value={somevalue} onChange={eventhandler}/>`. What you get back when you set an `onChange` prop on `SwitchBox` or `InputNumber` is the "cleaned up" value. Not an `event` object. **The `Input` component is different.** There you have to "convert" the `event` object to get to the actual input value. 

Had I instead opted for something like this:
```javascript
<Input onChange={this.handleNameChange}/>
```
then, the `handleNameChange` method would have to be different from all the other `handle*Change` methods there. I.e.:

```javascript
handleNameChange = event => {
  const name = event.target.value;
  const { id, type, value } = this.props;
  this.props.onRowChange({
    id,
    name,
    type,
    value,
  });
};
```

So there is consistency in the handler methods but some inconsistency in the `onChange` props usage.

Reply:
To me, this looks clearer than what we have now and I think we should go with it:
```js
handleNameChange = event => {
  const name = event.target.value;
  const { id, type, value } = this.props;
  this.props.onRowChange({
    id,
    name,
    type,
    value,
  });
};
```

Reply:
Thing is, if you do this, all the handlers that were quite "pure" before are now mixed. Some take an `event`, some take an actual value from the controlled antd component. See [this list of methods](https://github.com/mozilla/delivery-console/blob/f181fd8050b540ff22468a1e2f74616a7672a3e7/src/workflows/recipes/components/PreferenceRolloutFields.js#L192-L230) now. More mixed bag. 

But I don't care enough. I removed all the discussed inline functions so it's `<Input onChange={this.handlerFunction}/>` etc.